exiftool/exiftool.py

```diff
@@ -284,96 +225,379 @@ class ExifTool(object):
+		self._no_output = None # TODO examine whether this is needed
+		self._logger = None
+
+
+
+
+		# --- set variables via properties (which do the error checking) --
+
+		# set first, so that debug and info messages get logged
+		self.logger = logger
+
+		# use the passed in parameter, or the default if not set
+		# error checking is done in the property.setter
+		self.executable = executable if executable is not None else constants.DEFAULT_EXECUTABLE
+		self.common_args = common_args
+
+		# set the property, error checking happens in the property.setter
+		self.config_file = config_file
+
+
+
+
+
+
+		# --- run any remaining initialization code ---
+
+		random.seed(None) # initialize random number generator
+
+
+
+
+	#######################################################################################
+	#################################### MAGIC METHODS ####################################
+	#######################################################################################
+
+	# ----------------------------------------------------------------------------------------------------------------------
+
+	def __enter__(self):
+		self.run()
+		return self
+
+	# ----------------------------------------------------------------------------------------------------------------------
+	def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+		if self.running:
+			self.terminate()
+
+	# ----------------------------------------------------------------------------------------------------------------------
+	def __del__(self) -> None:
+		if self.running:
+			# indicate that __del__ has been started - allows running alternate code path in terminate()
+			self.terminate(_del=True)
+
+
+
+
+	########################################################################################
+	#################################### PROPERTIES R/w ####################################
+	########################################################################################
+
+	# ----------------------------------------------------------------------------------------------------------------------
+
+	@property
+	def executable(self):
+		return self._executable
+
+	@executable.setter
+	def executable(self, new_executable) -> None:
+		"""
+		Set the executable.  Does error checking.
+		
+		in testing, shutil.which() will work if a complete path is given, but this isn't clear, so we explicitly check and don't search if path exists
+		"""
+		# cannot set executable when process is running
+		if self.running:
+			raise RuntimeError( 'Cannot set new executable while Exiftool is running' )
+
+		abs_path: Optional[str] = None
 		
-		# error checking
-		if find_executable(self.executable) is None:
-			raise FileNotFoundError( '"{}" is not found, on path or as absolute path'.format(self.executable) )
+		if Path(new_executable).exists():
+			abs_path = new_executable
+		else:
+			# Python 3.3+ required
+			abs_path = shutil.which(new_executable)
```

Comment:
Interesting take on it.  Certainly there has to be error checking and error throwing before/during the `.run()` call, but in what use case would someone set an invalid executable, then wait till it's available before calling `.run()`?

Wouldn't "not checking" for errors on setting the property lead to latent bugs?  Like a dev:
1. sets a bad path
2. then goes to run it later to see the error in `.run()`, which currently would crash at one of the error handlers, probably line 603
3. to then have to try to traceback where the error came about... 
4. to the fact that they called the constructor with a bad path or set it in some random place using the property?

Reply:
Sorry, I didn't get a notification for your responses, so just saw these. 

In general, I consider checking that a file exists as being the responsibility of application code and not library code. Obviously, good library code should throw an error if it tries to access a file that doesn't exist, but until that access, a developer can have lots of valid application reasons for providing an "invalid" path. I might want to configure my class to only use the exe in a known location, but download it if it's not currently present. Downloading before constructing seems as valid constructing before downloading to me, in that circumstance. The library shouldn't care either way until it actually tries to call the path. The situation you describe certainly makes tracing the bug more difficult, but I'm not sure how much value it adds for a developer.

Reply:
lol.  Probably then, a possible course of action is to have a flag that changes the default behavior to have it both ways.

I think that a library should be just as robust so that a lazy programmer can't shoot themselves in the foot.  

> I might want to configure my class to only use the exe in a known location, but download it if it's not currently present. 
In the case that the executable doesn't exist, currently if you attempted to construct the object, it would raise an error in the constructor (well technically when the constructor sets `self.executable = ` it then errors out on line 351 inside the property).  

This would indicate to the program using ExifTool that the executable does not exist, and should then run the download routine and then reconstruct the object.

I went and patched some bad executable to the code and it dumps stuff out like this
```python
import exiftool
a = exiftool.ExifTool()
a.run()
```
<pre>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\_Digivents_Suite-v2_\script\lib\python\pyexiftool\exiftool\exiftool.py", line 619, in run
    raise fnfe
  File "D:\_Digivents_Suite-v2_\script\lib\python\pyexiftool\exiftool\exiftool.py", line 612, in run
    self._process = subprocess.Popen(
  File "D:\Utils\Dev_Tools\Python\Python39\lib\subprocess.py", line 947, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "D:\Utils\Dev_Tools\Python\Python39\lib\subprocess.py", line 1416, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
</pre>

The calling programmer would have to look at the `subprocess.Popen()` and then trace back the `proc_args` construction to figure out that it was the executable.  Then they have to go back and see when exactly they set the `executable` in their code path... could be multiple places, and do a trace of which set caused the problems... 

Anyways, are you using it in your code assuming it won't check?  I think the original code checked too, albeit not as robustly as the property does

Reply:
lol.  Probably then, a possible course of action is to have a flag that changes the default behavior to have it both ways.

I think that a library should be just as robust so that a lazy programmer can't shoot themselves in the foot.  

> I might want to configure my class to only use the exe in a known location, but download it if it's not currently present. 

In the case that the executable doesn't exist, currently if you attempted to construct the object, it would raise an error in the constructor (well technically when the constructor sets `self.executable =` it then errors out on line 351 inside the property).  

This would indicate to the program using ExifTool that the executable does not exist, and should then run the download routine and then reconstruct the object.

I went and patched some bad executable to the code and it dumps stuff out like this
```python
import exiftool
a = exiftool.ExifTool()
a.run()
```
<pre>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\TEMP\pyexiftool\exiftool\exiftool.py", line 619, in run
    raise fnfe
  File "D:\TEMP\pyexiftool\exiftool\exiftool.py", line 612, in run
    self._process = subprocess.Popen(
  File "D:\Utils\Python39\lib\subprocess.py", line 947, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "D:\Utils\Python39\lib\subprocess.py", line 1416, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
</pre>

The calling programmer would have to look at the `subprocess.Popen()` and then trace back the `proc_args` construction to figure out that it was the executable.  Then they have to go back and see when exactly they set the `executable` in their code path... could be multiple places, and do a trace of which set caused the problems... 

Anyways, are you using it in your code assuming it won't check?  I think the original code checked too, albeit not as robustly as the property does

Reply:
Yes, you could do:
```python
try:
    a = ExifTool('/my/bad/path')
except FileNotFoundError:
    download_exiftool()
    a = ExifTool('/my/bad/path')
```
In a GUI, though, `download_exiftool()` is long running (depending on bandwidth) and probably needs to be run asynchronously. It would be terribly convenient to have a preconfigured `ExifTool` instance pointing at the path where the executable _will_ be, and let the app developer do whatever they need to make sure the download has actually worked and the executable is available.

But even this is a contrived example. I just think that a file not existing isn't really an error until you actually try to access that file. I could delete and redownload the exiftool executable a hundred times after I construct an `ExifTool` object, and it wouldn't make a bit of difference.

And yeah, the old library does check for the executable on construction, which I don't love. I currently do a check on application launch so the user gets a message if ExifTool isn't found in `PATH`, and I'm intending a feature to let the user manually locate the executable.

As a side note, I think that exception reports the bad filename:
```python
try:
    ...
except FileNotFoundError as e:
    print(e.filename)
```

Anyway, this is also just a difference in library style preference. I would find it very convenient to have a way to disable the checking code.